src/disk_cache.rs

```diff
@@ -0,0 +1,1064 @@
+    Inconsistent { start: usize, end: usize },
+}
+
+pub trait LocalCacheStorage: Send + Sync + std::fmt::Debug + Display + 'static {
+    fn entry(
+        &self,
+        location: &object_store::path::Path,
+        part_size: usize,
+    ) -> Box<dyn LocalCacheEntry>;
+}
+
+#[async_trait::async_trait]
+pub trait LocalCacheEntry: Send + Sync + std::fmt::Debug + 'static {
+    async fn save_part(&self, part_number: PartID, buf: Bytes) -> object_store::Result<()>;
+
+    async fn read_part(&self, part_number: PartID) -> object_store::Result<Option<Bytes>>;
+
+    /// might be useful on rewriting GET request on the prefetch phase. the cached files are
+    /// expected to be in the same folder, so it'd be expected to be fast without expensive
+    /// globbing.
+    #[cfg(test)]
+    async fn cached_parts(&self) -> object_store::Result<Vec<PartID>>;
+
+    async fn save_head(&self, meta: (&ObjectMeta, &Attributes)) -> object_store::Result<()>;
+
+    async fn read_head(&self) -> object_store::Result<Option<(ObjectMeta, Attributes)>>;
+}
+
+#[derive(Debug)]
+struct DiskCacheStorage {
+    root_folder: std::path::PathBuf,
+}
+
+impl LocalCacheStorage for DiskCacheStorage {
+    fn entry(
+        &self,
+        location: &object_store::path::Path,
+        part_size: usize,
+    ) -> Box<dyn LocalCacheEntry> {
+        Box::new(DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size,
+        })
+    }
+}
+
+impl Display for DiskCacheStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "DiskCacheStorage({})", self.root_folder.display())
+    }
+}
+
+#[derive(Debug)]
+struct DiskCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: Path,
+    part_size: usize,
+}
+
+impl DiskCacheEntry {
+    async fn atomic_write(&self, path: &std::path::Path, buf: Bytes) -> object_store::Result<()> {
+        let tmp_path = path.with_extension(format!("_tmp{}", self.make_rand_suffix()));
+
+        // ensure the parent folder exists
+        if let Some(folder_path) = tmp_path.parent() {
+            fs::create_dir_all(folder_path).await.map_err(wrap_io_err)?;
+        }
+
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&tmp_path)
+            .await
+            .map_err(wrap_io_err)?;
+        file.write_all(&buf).await.map_err(wrap_io_err)?;
+        fs::rename(tmp_path, path).await.map_err(wrap_io_err)
+    }
+
+    fn make_part_path(
+        root_folder: std::path::PathBuf,
+        location: &Path,
+        part_number: usize,
+        part_size: usize,
+    ) -> std::path::PathBuf {
+        // containing the part size in the file name, allows user change the part size on
+        // the fly, without the need to invalidate the cache.
+        let part_size_name = if part_size % (1024 * 1024) == 0 {
```

Comment:
example in the comment would make it more readable.

Reply:
added example in 46c32c5